impl/maven-core/src/test/java/org/apache/maven/model/WrapperPropertiesTest.java

```diff
@@ -0,0 +1,264 @@
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class WrapperPropertiesTest {
+    private Map<String, String> backingStore;
+    private AtomicBoolean getterCalled;
+    private AtomicBoolean setterCalled;
+    private WrapperProperties wrapper;
+
+    @BeforeEach
+    void setUp() {
+        backingStore = new HashMap<>();
+        getterCalled = new AtomicBoolean(false);
+        setterCalled = new AtomicBoolean(false);
+
+        Supplier<Map<String, String>> getter = () -> {
+            getterCalled.set(true);
+            return new HashMap<>(backingStore);
+        };
+
+        Consumer<Properties> setter = props -> {
+            setterCalled.set(true);
+            backingStore.clear();
+            props.forEach((k, v) -> backingStore.put(k.toString(), v.toString()));
+        };
+
+        wrapper = new WrapperProperties(getter, setter);
+    }
+
+    @Test
+    void testInitialization() {
+        wrapper.getProperty("any");
+        assertTrue(getterCalled.get());
+    }
+
+    @Test
+    void testBasicOperations() {
+        // Test set and get
+        wrapper.setProperty("key1", "value1");
+        assertEquals("value1", wrapper.getProperty("key1"));
+        assertEquals("value1", backingStore.get("key1"));
+        assertTrue(setterCalled.get(), "Setter should be called after modification");
+
+        // Test remove
+        wrapper.remove("key1");
+        assertNull(wrapper.getProperty("key1"));
+        assertFalse(backingStore.containsKey("key1"));
+
+        // Test contains
+        wrapper.setProperty("key2", "value2");
+        assertTrue(wrapper.containsKey("key2"));
+        assertTrue(wrapper.containsValue("value2"));
+    }
+
+    @Test
+    void testOrderPreservation() {
+        wrapper.setProperty("a", "1");
+        wrapper.setProperty("b", "2");
+        wrapper.setProperty("c", "3");
+
+        // Check iteration order matches insertion order
+        Set<Object> keys = wrapper.keySet();
+        Object[] keyArray = keys.toArray();
+        assertEquals("a", keyArray[0]);
+        assertEquals("b", keyArray[1]);
+        assertEquals("c", keyArray[2]);
+
+        // Modify existing value and check order remains
+        wrapper.setProperty("b", "22");
+        keyArray = wrapper.keySet().toArray();
+        assertEquals("a", keyArray[0]);
+        assertEquals("b", keyArray[1]);
+        assertEquals("c", keyArray[2]);
+    }
+
+    @Test
+    @Disabled("should not throw NPE but does...")
```

Comment:
Is this a bug? If so, can you file an issue for that?

Reply:
lets dedicated this 

- https://github.com/apache/maven/pull/2437